### PR TITLE
Brownfield essentials dagster-external protoype

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Internal companion PR: https://github.com/dagster-io/internal/pull/6440

Prototype implementation of `dagster-external` library used for separation of orchestration and transform layers.

New resource `ExternalExecutionResource` works similarly to `DbtCliResource`. Create an asset with this resource and you can call `run` on the resource to execute an arbitrary program (via `Popen`). The resource pattern allows for the user to pull arbitrary (JSON-serializable) data off the `OpExecutionContext` and `DagsterInstance`, which is packaged together with other standard info (e.g. `run_id`) and sent over the wire to be reconstructed as an `ExternalExecutionContext` on the other side.

The external process should import `dagster_external` and call `init_dagster_external` (as of this PR, only Python is supported but use of `dagster-external` as a CLI is coming soon). This will read env vars injected by the spawning orchestration process, which are used to establish communications with the orchestration process. As of this PR stdio, a user-provided FIFO, or an automatically created temporary FIFO are IPC options (they can be varied independently for input and output to the external process). Socket support is planned.

`init_dagster_external` returns a singleton instance of `ExternalExecutionContext`-- after initialization, the singleton can also be retrieved by calling `ExternalExecutionContext.get()`.

The `ExternalExecutionContext` has a variety of accessors for the serialized context information sent from the orchestration process, as well as methods to stream data back to the orchestration process. This PR supports only a small number of such "reporter" methods and the names are not set in stone, this will be iterated on in the near future.

## How I Tested These Changes

New unit tests.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15759** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15759" target="_blank">Graphite</a> 
    * **PR #15760** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15760" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/15760?utm_source=stack-comment).

File: python_modules/dagster-toolkit/dagster_toolkit/protocol.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+class OTPExecutionContext(TypedDict):
+    asset_key: str
+    code_version: Optional[str]
+    data_provenance: Optional["OTPDataProvenance"]
+    partition_key: str
+    partition_key_range: Optional["OTPPartitionKeyRange"]
+    partition_time_window: Optional["OTPTimeWindow"]
+    run_id: str
+    tags: Mapping[str, str]
+    op_config: Any
```
Comment (schrockn) [inline python_modules/dagster-toolkit/dagster_toolkit/protocol.py:42]: op_config should be opt-in

File: pyright/master/requirements.txt [hunk 2]
```diff
@@ -31,6 +31,7 @@
 -e python_modules/dagit
 -e python_modules/dagster-graphql
 -e python_modules/dagster-test
+-e python_modules/dagster-toolkit
```
Comment (schrockn) [inline pyright/master/requirements.txt:34]: rename to dagster-extenral

File: python_modules/dagster-toolkit/dagster_toolkit/__init__.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from dagster_toolkit.protocol import (
+    OTPDataProvenance as OTPDataProvenance,
+    OTPExecutionContext as OTPExecutionContext,
+    OTPPartitionKeyRange as OTPPartitionKeyRange,
+    OTPTimeWindow as OTPTimeWindow,
+)
+from dagster_toolkit.session import Session as Session
```
Comment (schrockn) [inline python_modules/dagster-toolkit/dagster_toolkit/__init__.py:7]: remove Session and instead do init_dagster_external

File: python_modules/dagster-toolkit/dagster_toolkit/session.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    def __init__(self) -> None:
+        self._channel_type = os.getenv(DAGSTER_TOOLKIT_ENV_KEYS["channel_type"], "stdio")
+        if self._channel_type == "stdio":
+            # print("STDIO CHANNEL")
+            # print("READING")
+            # print(self._input_stream.read())
+            self._input_stream = sys.stdin
+            self._output_stream = sys.stdout
+        elif self._channel_type == "fifo":
+            self._input_stream = open(_safe_getenv(DAGSTER_TOOLKIT_ENV_KEYS["input"]), "r")
```
Comment (schrockn) [inline python_modules/dagster-toolkit/dagster_toolkit/session.py:51]: nit: do not call os.getenv in `__init__`

**Review by schrockn — DISMISSED** [review]

File: python_modules/dagster-external/dagster_external/context.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        return self._data["partition_time_window"]
+
+    @property
+    def run_id(self) -> str:
+        return self._data["run_id"]
+
+    @property
+    def tags(self) -> Mapping[str, str]:
+        return self._data["tags"]
+
```
Comment (alangenfeld) [inline python_modules/dagster-external/dagster_external/context.py:110]: nit: run tags vs op tags , not sure its safe to count on this only being used in assets or assets never adding tags to propagate to their op

File: python_modules/dagster/dagster/_core/external_execution/context.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        data_provenance=(_convert_data_provenance(asset_provenance) if asset_provenance else None),
+        partition_key=partition_key,
+        partition_key_range=(
+            _convert_partition_key_range(partition_key_range) if partition_key_range else None
+        ),
+        partition_time_window=(
+            _convert_time_window(partition_time_window) if partition_time_window else None
+        ),
+        run_id=context.run_id,
+        run_tags=context.run_tags,
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/external_execution/context.py:38]: on further thought - I think tags is sort of like config in that we can leave it to the user to pass through what they care about as userdata and omit it from the default set

**Review by alangenfeld — APPROVED** [review]
if @schrockn has feedback you can incorporate in to a follow-up PR

File: python_modules/dagster/dagster/_core/external_execution/context.py (hunk 6 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/external_execution/context.py:38]: ok I'll address this in the followup

**Pull request merged**